Improve ValidateLength error message consistency and refactor validation tests

jorgeasaurus opened the pull request:
## PR Summary
This PR improves the consistency of ValidateLength error messages and significantly enhances the quality of validation attribute tests. The changes improve both user experience (clearer error messages) and code maintainability (more reliable, readable tests).

Key improvements include:

### Error Message Consistency
1. Updated `ValidateLengthMaxLengthFailure` resource string in Metadata.resx to match the format of `ValidateLengthMinLengthFailure`
2. Changed format from `"The character length of the {1} argument is too long..."` to `"The character length ({1}) of the argument is too long..."`
3. Both min and max length errors now use consistent language and formatting with actual length in parentheses

### Test Quality Improvements
1. Replaced generic function names (`foo`) with descriptive PowerShell-compliant names (`Test-ArrayCount`, `Test-NumericRange`, `Test-StringLength`)
2. Replaced generic parameter names (`bar`) with meaningful names (`$Items`, `$Number`, `$InputString`)
3. Updated error assertions to use `-PassThru` with captured error variables instead of relying on `$error[0]`
4. Added comprehensive ValidateLength test context with specific error message format validation

## PR Context
The ValidateLength error messages were inconsistent in format:
- **MinLength**: "The character length (3) of the argument is too short..."
- **MaxLength** (old): "The character length of the 5 argument is too long..." ❌

The updated format makes both messages consistent:
- **MinLength**: "The character length (3) of the argument is too short. Specify an argument with a length that is greater than or equal to "5"..."
- **MaxLength** (new): "The character length (5) of the argument is too long. Specify an argument with a length that is shorter than or equal to "2"..." ✅

Additionally, the test refactoring addresses best practices by using `-PassThru` to capture exceptions directly rather than reading from the global `$error` array, making tests more reliable and less prone to interference from other test runs.

## PR Checklist
- [x] PR has a meaningful title
  - Use the present tense and imperative mood when describing your changes
- [x] Summarized changes
- [x] Make sure all .h, .cpp, .cs, .ps1 and .psm1 files have the correct copyright header
- [x] This PR is ready to merge. If this PR is a work in progress, please open this as a Draft Pull Request and mark it as Ready to Review when it is ready to merge.

## Breaking changes
- [ ] None
OR
- [ ] Experimental feature(s) needed
  - Experimental feature name(s):

## User-facing changes
- [x] Not Applicable
OR
- [ ] Documentation needed
  - Issue filed:

## Testing - New and feature
- [x] N/A or can only be tested interactively
OR
- [ ] Make sure you've added a new test if existing tests do not effectively test the code changed

jorgeasaurus committed:
Fix ValidateLength error message formatting issue
  (#25566)

  ## Problem
  The ValidateLengthMaxLengthFailure error message was
  displaying incorrectly,
  showing the length value where the parameter name
  should appear:
  - Actual: "The character length of the 5 argument is
  too long..."
  - Expected: "The character length of the "argument"
  argument is too long..."

  This occurred because the resource string was updated
   to expect three parameters
  (MaxLength, actual length, and parameter name) but
  the code was only passing two
  parameters, causing the format string to use the
  length value in place of the
  parameter name.

  ## Root Cause
  The resource string format was changed to include a
  parameter name placeholder:
  - Old format: "The character length of the {1}
  argument..."
  - New format: "The character length of the "{2}"
  argument is too long at {1} characters..."

  But the corresponding code change to pass the
  parameter name was not implemented.

  ## Solution
  1. Updated ValidateLengthAttribute.ValidateElement()
  to pass "argument" as the
     third parameter when throwing
  ValidationMetadataException for max length
     validation failures.

  2. Updated the ValidateLengthMaxLengthFailure
  resource string in Metadata.resx
     to properly format the error message with all
  three placeholders.

  While the current architecture doesn't support
  passing actual parameter names
  to validation attributes (which would require
  significant architectural changes),
  using "argument" as a generic placeholder prevents
  the confusing formatting
  error and provides a clear, consistent error message.

  ## Changes
  - Modified
  src/System.Management.Automation/engine/Attributes.cs
   to pass
    "argument" as third parameter in ValidateLength max
   length validation
  - Updated src/System.Management.Automation/resources/
  Metadata.resx to include
    proper formatting for parameter name placeholder
  - Added comprehensive tests in test/powershell/engine
  /Basic/ValidateAttributes.Tests.ps1
    to verify the fix

  ## Testing
  Added tests to verify:
  - Maximum length validation error formatting shows
  "argument" placeholder
  - Minimum length validation remains unchanged
  (different format)
  - Non-string parameter validation works correctly
  - Valid string lengths pass validation
  - Error message contains actual parameter name in
  outer exception

  The fix ensures error messages are properly formatted
   and readable while
  maintaining backward compatibility.

  Fixes #25566

jorgeasaurus committed:
Merge branch 'master' into master

jorgeasaurus committed:
Merge branch 'master' into master

iSazonov commented on src/System.Management.Automation/resources/Metadata.resx:
```diff
@@ -175,7 +175,7 @@
   </data>
   <data name="ValidateLengthMaxLengthFailure" xml:space="preserve">
-    <value>The character length of the {1} argument is too long. Shorten the character length of the argument so it is fewer than or equal to "{0}" characters, and then try the command again.</value>
+    <value>The character length of the "{2}" argument is too long at {1} characters. Shorten the character length so it is fewer than or equal to {0} characters.</value>
```
This should be like `ValidateLengthMinLengthFailure`.

jorgeasaurus committed:
Merge branch 'PowerShell:master' into master

jorgeasaurus committed:
Refactor error message for ValidateLength to improve clarity and consistency

jorgeasaurus commented:
Thank you @iSazonov for the review feedback. I've updated the PR to make `ValidateLengthMaxLengthFailure` consistent with `ValidateLengthMinLengthFailure` format.

### Changes:

**1. Error Message Format** ([Metadata.resx:178](src/System.Management.Automation/resources/Metadata.resx#L178))

Changed from:
```
The character length of the "{2}" argument is too long at {1} characters. Shorten the character length so it is fewer than or equal to {0} characters.
```

To:
```
The character length ({1}) of the argument is too long. Specify an argument with a length that is shorter than or equal to "{0}", and then try the command again.
```

**2. Code Update** ([Attributes.cs:872](src/System.Management.Automation/engine/Attributes.cs#L872))

Removed the third parameter to match MinLength's two-parameter format:
```csharp
// Before
throw new ValidationMetadataException(
    "ValidateLengthMaxLengthFailure",
    null,
    Metadata.ValidateLengthMaxLengthFailure,
    MaxLength, len, "argument");

// After
throw new ValidationMetadataException(
    "ValidateLengthMaxLengthFailure",
    null,
    Metadata.ValidateLengthMaxLengthFailure,
    MaxLength, len);
```

**3. Test Update** ([ValidateAttributes.Tests.ps1:319-320](test/powershell/engine/Basic/ValidateAttributes.Tests.ps1#L319))

Updated test to verify the new consistent format:
```powershell
# Check the inner exception message is properly formatted and consistent with MinLength format
$error[0].Exception.InnerException.Message | Should -Match 'The character length \(5\) of the argument is too long\.'
$error[0].Exception.InnerException.Message | Should -Match 'shorter than or equal to "2"'
```

### Result:

Both error messages now follow the same consistent pattern:

**MinLength:**
```
The character length (3) of the argument is too short. Specify an argument with a length that is greater than or equal to "5", and then try the command again.
```

**MaxLength:**
```
The character length (6) of the argument is too long. Specify an argument with a length that is shorter than or equal to "2", and then try the command again.
```

jorgeasaurus committed:
Refactor ValidateLength test to improve error message formatting and consistency

jorgeasaurus committed:
Refactor ValidateLength tests to improve error handling and message formatting

jorgeasaurus committed:
Merge branch 'PowerShell:master' into master

jorgeasaurus committed:
Refactor ValidateAttributes.Tests to use descriptive function names for better clarity
